apps/els_lsp/src/els_bsp_provider.erl

```diff
@@ -0,0 +1,262 @@
+-include_lib("kernel/include/logger.hrl").
+
+%%==============================================================================
+%% Types
+%%==============================================================================
+-type state() :: #{ running          := boolean()         % BSP server running?
+                  , root_uri         := uri() | undefined % the root uri
+                  , pending_requests := list()            % pending requests
+                  , pending_sources  := list()            % pending source info
+                  }.
+-type method() :: binary().
+-type params() :: map() | null.
+-type request_response() :: {reply, any()} | {error, any()}.
+-type request_id() :: {reference(), reference(), atom() | pid()}.
+-type from() :: {pid(), reference()} | self.
+-type config() :: map().
+-type info_item() :: is_running.
+
+%%==============================================================================
+%% API
+%%==============================================================================
+-spec start(uri()) -> {ok, config()} | {error, term()}.
+start(RootUri) ->
+  els_provider:handle_request(?MODULE, {start, #{ root => RootUri }}).
+
+-spec maybe_start(uri()) -> {ok, config()} | {error, term()} | disabled.
+maybe_start(RootUri) ->
+  case els_config:get(bsp_enabled) of
+    false ->
+      disabled;
+    X when X =:= true orelse X =:= auto ->
+      start(RootUri)
+  end.
+
+-spec info(info_item()) -> any().
+info(Item) ->
+  case els_provider:handle_request(?MODULE, {info, #{ item => Item}}) of
+    {ok, Result} ->
+      Result;
+    {error, badarg} ->
+      erlang:error(badarg, [Item])
+  end.
+
+-spec request(method(), params()) -> request_response().
+request(Method, Params) ->
+  RequestId = send_request(Method, Params),
+  wait_response(RequestId, infinity).
+
+-spec send_request(method(), params()) -> request_id().
+send_request(Method, Params) ->
+  Mon = erlang:monitor(process, ?MODULE),
+  Ref = erlang:make_ref(),
+  From = {self(), Ref},
+  Request = #{ from => From, method => Method, params => Params },
+  ok = els_provider:handle_request(?MODULE, {send_request, Request}),
+  {Ref, Mon, ?MODULE}.
+
+-spec wait_response(request_id(), timeout()) -> request_response() | timeout.
+wait_response({Ref, Mon, ServerRef}, Timeout) ->
+  receive
+    {Ref, Response} ->
+      erlang:demonitor(Mon, [flush]),
+      Response;
+    {'DOWN', Mon,  _Type, _Object, Info} ->
+      erlang:demonitor(Mon, [flush]),
+      {error, {Info, ServerRef}}
+  after Timeout ->
+      timeout
+  end.
+
+%%==============================================================================
+%% els_provider functions
+%%==============================================================================
+-spec init() -> state().
+init() ->
+  #{ running          => false
+   , root_uri         => undefined
+   , pending_requests => []
+   , pending_sources  => [ apps_paths, deps_paths ]
+   }.
+
+-spec is_enabled() -> true.
+is_enabled() -> true.
```

Comment:
Should we have this disabled by default?

Reply:
The provider being enabled doesn't mean that the build server is started.